Handle Redirects within a page (e.g. root redirects onto an action)

lbradley-mm opened the pull request:
This also uses the main `Director::handleRequest()` so ALL routes are handled correctly
Changes the `Versioned::set_reading_mode()` to `Versioned::set_default_reading_mode()` as is not reset on the `Director::handleRequest()`

lbradley-mm committed:
Handle Redirects within a page (e.g. root redirects onto an action)

lbradley-mm committed:
Update to use `Versioned::set_default_reading_mode()`

Because the previous change uses the main `Director::handleRequest()` that resets the stage to Live

emteknetnz commented on src/Controllers/ShareDraftController.php:
```diff
@@ -120,21 +113,32 @@ public function preview(HTTPRequest $request)
+        $response = Director::singleton()->handleRequest($pageRequest);
+        if ($response->isRedirect()) {
+            // The redirect will probably be Absolute URL so just want the path
+            $newUrl = parse_url($response->getHeader('location'), PHP_URL_PATH);
```
what happens if the redirect is to a different website?

lbradley-mm replied:
I'll do some tests and find out

emteknetnz replied:
I've tested this locally, $newUrl will end up null so what'll happen is the share draft link will simply render the homepage.  Given redirector pages are all broken right, have it only broken for external links isn't a big deal.  Also I'm not sure what the UX should even look like since we probably wouldn't want to preview the external site.

emteknetnz commented on src/Controllers/ShareDraftController.php:
```diff
@@ -79,28 +81,19 @@ public function preview(HTTPRequest $request)
             // Temporarily un-secure the draft site and switch to draft
             $oldSecured = $this->getIsDraftSecured($session);
-            $oldMode = Versioned::get_reading_mode();
+            $oldMode = Versioned::get_default_reading_mode();
```
I'm not sure why you're switching `reading_mode()` to `default_reading_mode()`.  I've had a quick look through the codebase for Director::handleRequest() though I cannot find any reference to reading_mode that would make a difference?  Can you point to some code as to why this is necessary?

lbradley-mm replied:
Because I updated to use `Director::handleRequest()` (this resets the `reading_mode` but not the `default_reading_mode`)

emteknetnz replied:
Yeah I search for this in Director.php and some other related files, though I wasn't able to pinpoint where this actually happens - would you be able to point me in the right direction?

lbradley-mm replied:
The "Live" mode is being set because the `VersionedHTTPMiddleware::process` will apply it based on the `HTTPRequest` (which will not have the `stage=Stage` parameter) is being processed. See `Versioned::choose_site_stage()` which is called

emteknetnz replied:
OK that makes sense. I've added a suggested comment below to explain this to future developers

lbradley-mm replied:
added your comment

emteknetnz reviewed (changes requested):
Seems generally OK, through I've asked a couple of questions

emteknetnz commented on src/Controllers/ShareDraftController.php:
```diff
@@ -110,7 +103,7 @@ public function preview(HTTPRequest $request)
             } finally {
                 $this->setIsDraftSecured($session, $oldSecured);
-                Versioned::set_reading_mode($oldMode);
+                Versioned::set_default_reading_mode($oldMode);
```
```suggestion
                // Use set_default_reading_mode() instead of set_reading_mode() because that's
                // what's used in Versioned::choose_site_stage()
                Versioned::set_default_reading_mode($oldMode);
```

emteknetnz reviewed (changes requested):
Looks good, I've tested locally and it seems to work correctly.  Just one minor code suggestion I've made to add a comment about set_default_reading_mode

lbradley-mm committed:
Adds explanatory comment

Because the internal redirect doesn't pass in `?stage=Stage` we need to make sure the default reading mode is Stage

Co-authored-by: Steve Boyd <emteknetnz@gmail.com>
